File: packages/marketdata/src/l2book.ts
import type { EngineEvent, Side } from '@exchange/common';

/**
 * L2 (level-2) book aggregator.
 *
 * Unlike the engine's book — which knows per-order — the L2 view stores
 * only the TOTAL QTY at each price level on each side. That's what
 * traders typically see in an exchange's depth chart.
 *
 * Update rules:
 *   - OrderAccepted: pre-add the full qty at the order's (side, price).
 *     For MARKET orders (price = 0) we skip — they cannot rest.
 *   - Trade: reduce BOTH sides — the maker's level by qty (the maker is
 *     resting on the book) AND the taker's level by qty (we pre-added
 *     it on OrderAccepted, so we need to take back the part that just
 *     matched, leaving only the unfilled remainder visible at the
 *     taker's limit price).
 *   - OrderCanceled: remove the order's REMAINING qty from its level.
 *     Works for both user-cancels (resting orders) and terminal cancels
 *     (IOC remainder, STP, MARKET_NO_LIQUIDITY).
 *
 * This makes the L2 reflect the user's order the instant the engine
 * accepts it — no waiting for a "next submission" boundary.
 */

export interface L2Delta {
  side: Side;
  price: string;   // bigint as string for JSON
  qty: string;     // total qty at this level AFTER the change; '0' means remove
}

export interface L2Snapshot {
  bids: [string, string][];  // [price, qty]
  asks: [string, string][];
}

interface TrackedOrder {
  side: Side;
  price: bigint;          // 0n for MARKET (never adds to a level)
  remaining: bigint;
}

export class L2Book {
  private readonly bids = new Map<bigint, bigint>();
  private readonly asks = new Map<bigint, bigint>();
  private readonly orders = new Map<string, TrackedOrder>();

  apply(ev: EngineEvent): L2Delta[] {
    switch (ev.kind) {
      case 'OrderAccepted': {
        const o = ev.order;
        this.orders.set(o.orderId, {
          side: o.side,
          price: o.price,
          remaining: o.qty,
        });
        // MARKET orders never rest. Don't put them in the depth view.
        if (o.type === 'MARKET' || o.price === 0n) return [];
        return [this.applyLevelDelta(o.side, o.price, o.qty)];
      }
      case 'OrderRejected':
        return [];
      case 'Trade': {
        const deltas: L2Delta[] = [];
        const maker = this.orders.get(ev.makerOrderId);
        if (maker) {
          maker.remaining -= ev.qty;
          if (maker.remaining <= 0n) this.orders.delete(ev.makerOrderId);
          deltas.push(this.applyLevelDelta(maker.side, maker.price, -ev.qty));
        }
        const taker = this.orders.get(ev.takerOrderId);
        if (taker) {
          taker.remaining -= ev.qty;
          if (taker.remaining <= 0n) this.orders.delete(ev.takerOrderId);
          // taker.price is 0 for MARKET — those were never added to a level.
          if (taker.price > 0n) {
            deltas.push(this.applyLevelDelta(taker.side, taker.price, -ev.qty));
          }
        }
        return deltas;
      }
      case 'OrderCanceled': {
        const existing = this.orders.get(ev.orderId);
        if (!existing) return [];
        this.orders.delete(ev.orderId);
        if (existing.price === 0n) return []; // MARKET, never added
        return [this.applyLevelDelta(existing.side, existing.price, -existing.remaining)];
      }
    }
  }

  snapshot(levels: number = 50): L2Snapshot {
    return {
      bids: topN(this.bids, 'desc', levels),
      asks: topN(this.asks, 'asc', levels),
    };
  }

  private applyLevelDelta(side: Side, price: bigint, deltaQty: bigint): L2Delta {
    const tree = side === 'buy' ? this.bids : this.asks;
    const current = tree.get(price) ?? 0n;
    const next = current + deltaQty;
    if (next <= 0n) tree.delete(price);
    else tree.set(price, next);
    return { side, price: price.toString(), qty: (next > 0n ? next : 0n).toString() };
  }
}

function topN(
  m: Map<bigint, bigint>,
  order: 'asc' | 'desc',
  n: number
): [string, string][] {
  const sorted = [...m.entries()].sort((a, b) =>
    order === 'asc'
      ? a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0
      : a[0] < b[0] ? 1 : a[0] > b[0] ? -1 : 0
  );
  return sorted.slice(0, n).map(([p, q]) => [p.toString(), q.toString()]);
}

File: packages-go/marketdata/l2book.go
package main

import (
	"sort"
	"strconv"

	"github.com/KaustubhPatange/exchange/common"
)

// L2Book is an L2 (level-2) book aggregator.
//
// Unlike the engine's book — which knows per-order — the L2 view stores
// only the TOTAL QTY at each price level on each side. That's what
// traders typically see in an exchange's depth chart.
//
// Update rules:
//   - OrderAccepted: pre-add the full qty at the order's (side, price).
//     For MARKET orders (price = 0) we skip — they cannot rest.
//   - Trade: reduce BOTH sides — the maker's level by qty (the maker is
//     resting on the book) AND the taker's level by qty (we pre-added
//     it on OrderAccepted, so we need to take back the part that just
//     matched, leaving only the unfilled remainder visible at the
//     taker's limit price).
//   - OrderCanceled: remove the order's REMAINING qty from its level.
//     Works for both user-cancels (resting orders) and terminal cancels
//     (IOC remainder, STP, MARKET_NO_LIQUIDITY).
//
// This makes the L2 reflect the user's order the instant the engine
// accepts it — no waiting for a "next submission" boundary.
type L2Book struct {
	bids   map[int64]int64
	asks   map[int64]int64
	orders map[string]*trackedOrder
}

type trackedOrder struct {
	side      common.Side
	price     int64 // 0 for MARKET (never adds to a level)
	remaining int64
}

type L2Delta struct {
	Side  common.Side `json:"side"`
	Price string      `json:"price"`
	Qty   string      `json:"qty"` // total qty at this level AFTER the change; '0' means remove
}

type L2Snapshot struct {
	Bids [][2]string `json:"bids"` // [price, qty]
	Asks [][2]string `json:"asks"`
}

func NewL2Book() *L2Book {
	return &L2Book{
		bids:   make(map[int64]int64),
		asks:   make(map[int64]int64),
		orders: make(map[string]*trackedOrder),
	}
}

func (b *L2Book) Apply(ev common.EngineEvent) []L2Delta {
	switch e := ev.(type) {
	case *common.OrderAcceptedEvent:
		o := e.Order
		b.orders[o.OrderID] = &trackedOrder{side: o.Side, price: o.Price, remaining: o.Qty}
		if o.Type == common.OrderTypeMarket || o.Price == 0 {
			return nil
		}
		return []L2Delta{b.applyLevelDelta(o.Side, o.Price, o.Qty)}

	case *common.OrderRejectedEvent:
		return nil

	case *common.TradeEvent:
		var deltas []L2Delta
		if maker, ok := b.orders[e.MakerOrderID]; ok {
			maker.remaining -= e.Qty
			if maker.remaining <= 0 {
				delete(b.orders, e.MakerOrderID)
			}
			deltas = append(deltas, b.applyLevelDelta(maker.side, maker.price, -e.Qty))
		}
		if taker, ok := b.orders[e.TakerOrderID]; ok {
			taker.remaining -= e.Qty
			if taker.remaining <= 0 {
				delete(b.orders, e.TakerOrderID)
			}
			// taker.price is 0 for MARKET — those were never added to a level.
			if taker.price > 0 {
				deltas = append(deltas, b.applyLevelDelta(taker.side, taker.price, -e.Qty))
			}
		}
		return deltas

	case *common.OrderCanceledEvent:
		existing, ok := b.orders[e.OrderID]
		if !ok {
			return nil
		}
		delete(b.orders, e.OrderID)
		if existing.price == 0 {
			return nil // MARKET, never added
		}
		return []L2Delta{b.applyLevelDelta(existing.side, existing.price, -existing.remaining)}
	}
	return nil
}

func (b *L2Book) Clear() {
	b.bids = make(map[int64]int64)
	b.asks = make(map[int64]int64)
	b.orders = make(map[string]*trackedOrder)
}

func (b *L2Book) Snapshot(levels int) L2Snapshot {
	if levels == 0 {
		levels = 50
	}
	return L2Snapshot{
		Bids: topN(b.bids, false, levels),
		Asks: topN(b.asks, true, levels),
	}
}

func (b *L2Book) applyLevelDelta(side common.Side, price int64, deltaQty int64) L2Delta {
	tree := b.asks
	if side == common.SideBuy {
		tree = b.bids
	}
	next := tree[price] + deltaQty
	if next <= 0 {
		delete(tree, price)
	} else {
		tree[price] = next
	}
	if next < 0 {
		next = 0
	}
	return L2Delta{Side: side, Price: strconv.FormatInt(price, 10), Qty: strconv.FormatInt(next, 10)}
}

func topN(m map[int64]int64, ascending bool, n int) [][2]string {
	prices := make([]int64, 0, len(m))
	for p := range m {
		prices = append(prices, p)
	}
	if ascending {
		sort.Slice(prices, func(i, j int) bool { return prices[i] < prices[j] })
	} else {
		sort.Slice(prices, func(i, j int) bool { return prices[i] > prices[j] })
	}
	if n < len(prices) {
		prices = prices[:n]
	}
	out := make([][2]string, len(prices))
	for i, p := range prices {
		out[i] = [2]string{strconv.FormatInt(p, 10), strconv.FormatInt(m[p], 10)}
	}
	return out
}
